src/models/mSeriesChallenge.ts

```diff
@@ -0,0 +1,268 @@
+import { ListMeta, Model, ModelList } from '../model'
+import { AuthenticatedResponse, SessionHandler } from '../session'
+import { MSeriesChallengeLeaderboardParticipants, MSeriesChallengeLeaderboardResponse, MSeriesChallengeParticipant, MSeriesChallengeParticipantListResponse, MSeriesChallengeParticipantResponse, MSeriesChallengeParticipants, MSeriesChallengeParticipantSorting } from './mSeriesChallengeParticipant'
+
+export const enum MSeriesChallengeType {
+  TimeBased = 'timeBased',
+  GoalBased = 'goalBased',
+}
+
+export const enum MSeriesChallengeFocus {
+  Points = 'points',
+  CaloricBurn = 'caloricBurn',
+  Distance = 'distance',
+  Duration = 'duration',
+}
+
+export const enum MSeriesChallengeRelationship {
+  Owned = 'owned',
+  Joined = 'joined'
+}
+
+export const enum MSeriesChallengeSorting {
+  ID = 'id',
+  StartAt = 'startAt',
+  EndAt = 'endAt',
+  CreatedAt = 'createdAt'
+}
+
+export interface MSeriesChallengeData {
+  joinCode: string | undefined
+  id: number
+  name: string
+  userId: number
+  isPublic: boolean
+  userLimit: number | null
+  challengeType: MSeriesChallengeType
+  startAt: string
+  endAt: string | null
+  focus: MSeriesChallengeFocus
+  goal: number
+  isCompleted: boolean
+}
+
+export interface MSeriesChallengeResponse extends AuthenticatedResponse {
+  mSeriesChallenge: MSeriesChallengeData
+}
+
+export interface MSeriesChallengeListResponse extends AuthenticatedResponse {
+  mSeriesChallenges: MSeriesChallengeData[]
+  mSeriesChallengesMeta: MSeriesChallengeListResponseMeta
+}
+
+export interface MSeriesChallengeListResponseMeta extends ListMeta {
+  from: string | undefined
+  to: string | undefined
+  isCompleted: boolean
+  relationship: MSeriesChallengeRelationship
+  sort: MSeriesChallengeSorting
+}
+
+export class MSeriesChallenges extends ModelList<
+MSeriesChallenge,
+MSeriesChallengeData,
+MSeriesChallengeListResponseMeta
+> {
+  constructor (mSeriesChallenges: MSeriesChallengeData[], mSeriesChallengesMeta: MSeriesChallengeListResponseMeta, sessionHandler: SessionHandler) {
+    super(MSeriesChallenge, mSeriesChallenges, mSeriesChallengesMeta, sessionHandler)
+  }
+}
+
+export class MSeriesChallenge extends Model {
+  private _mSeriesChallengeData: MSeriesChallengeData
+  private _mSeriesChallengeParticipant?: MSeriesChallengeParticipant
```

Comment:
Why is `mSeriesChallengeParticipant` a property of this model?

Reply:
It represents the current session user's relation to a challenge. If they have joined the challenge, it's their participant data.

Reply:
I understand, but it's association is unclear and very odd.  It feels like it should be nested under the `MSeriesChallenge` model as `MSeriesChallengeParticipants`.

Reply:
The 'challenge' views on the mSeries app almost always show the current session user's relation to the challenge(rank, etc). So it made sense to me to keep that data coupled to the challenge itself.

I understand it's confusing, but it seemed more confusing to me to instead store a list of `MSeriesChallengeParticipants` and have to **find** the current session user in that list to get their data.